### PR TITLE
Mgdapi 31 Updating the prepeare release script 

### PR DESCRIPTION
**Pull request by Boomatang**
# Description

The prepare release script now accepts more arguments with all the user to sepicfiy the `olm-catalog` resource be released. 
Also is the ability to set the quay.io org that will point to the catalogue images. This is useful when doing upgrade testing of bundles. Short readme added with a description of how to use the script and the arguments that can be set, plus the current defaults.

**JIRA** https://issues.redhat.com/browse/MGDAPI-31

**Requires https://github.com/integr8ly/delorean/pull/148 to be merged**

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist
- [ ] This change requires a documentation update 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added a test case that will be used to verify my changes 
- [ ] Verified independently on a cluster by reviewer

### CODE REVIEW

File: scripts/prepare-release.sh [hunk 1]
```diff
@@ -2,30 +2,51 @@
 set -e
 set -o pipefail
 
-PREVIOUS_VERSION=$(grep integreatly-operator deploy/olm-catalog/integreatly-operator/integreatly-operator.package.yaml | awk -F v '{print $2}')
+
+PREVIOUS_VERSION=$(grep $TYPE deploy/olm-catalog/$TYPE/$TYPE.package.yaml | awk -F v '{print $2}') || echo "No previous version"
+
+if [[ -z "$TYPE" ]]; then
```
Comment (philbrookes) [inline scripts/prepare-release.sh:8]: Doesn't this need to occur before line 6 which is reading this value?

**Comment by philbrookes** [comment]
e2e rate limit flake
/retest

**Comment by sergioifg94** [comment]
Just a couple of questions about this:

1. I see the version of the CSV is `0.0.0`, shouldn't it be `1.0.0` instead?
2. I see that the CSV name is `managed-3scale`. Are we sure we're going for that name for it, or `managed-api-service`? Maybe @philbrookes can clarify this one

Just confirming to be in sync with the manifests and metadata that I bring into managed tenants

**Comment by Boomatang** [comment]
@sergioifg94 These are good questions that I will try answer.

1. The version `0.0.0` was done as a place holder. I should have been clear on that in my description. This is not a release of the api managed service, that CSV will install the full RHMI. I would be expecting that CSV to be removed when a genuine release is been created.
2. I had spoke to @philbrookes about using the `managed-3scale` and at the time it seemed like the right name. As the initial `0.0.0` CSV is a place holder there will be no issue rerunning the script with the correct values "service name & version" when we are ready to do an official release.

**Comment by philbrookes** [comment]
1. `0.0.0` is fine for now.
2. This should be `managed-api` now, can we rename the directory, if we merge after https://github.com/integr8ly/integreatly-operator/pull/1175 then this PR can use `managed-api` as the install type Param too.

**Comment by Boomatang** [comment]
Thanks, I will keep an eye for #1175 to be merged and then make the changes to suit.

File: scripts/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+This script creates a release csv for the operator under the `olm-catalog` using `operator-sdk generate csv`
+
+**Required System Variables**
+
+- SEMVER -> valid x.y.z format. Usage: `SEMVER=2.6.0`
+
+**Optional System Variables**
+
+- TYPE -> Which resource the release csv will be created under. Current Default: `integreatly-operator`.  Usage: `TYPE=managed-api-service`
+- ORG -> The org to where any images maybe stored on quay.io. Setting this will change the affect the location images
```
Comment (obrienrobert) [inline scripts/README.md:12]: `the affect the`: does this need to be reworded?

Comment (Boomatang) [reply scripts/README.md:12]: Yes, yes it does. Even when I read it not it don't make must sense not.

Comment (Boomatang) [reply scripts/README.md]: @obrienrobert I have just push just push an update if you would like to review

File: scripts/README.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+- SEMVER -> valid x.y.z format. Usage: `SEMVER=2.6.0`
+
+**Optional System Variables**
+
+- TYPE -> Which resource the release csv will be created under. Current Default: `integreatly-operator`.  Usage: `TYPE=managed-api-service`
+- ORG -> The org to where any images maybe stored on quay.io. Setting the ORG will change the image locations specified in \<type>.\<version>.clusterserviceversion.yaml under `containerImage` and `image`. Default: `integreatly`. Usage: `ORG=<user-repo>`.
+
+**Usage**
+
+`SEMVER=2.6.0 TPYE=managed-api-service ORG=<user-repo> ./prepare-release.sh`
```
Comment (obrienrobert) [inline scripts/README.md:16]: Small one: `TPYE` -> `TYPE`.

File: deploy/olm-catalog/managed-api-service/managed-api.package-service.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+channels:
+- currentCSV: managed-api-service.v0.0.0
```
Comment (laurafitzgerald) [inline deploy/olm-catalog/managed-api-service/managed-api.package-service.yaml:2]: Interesting that v0.0.0 versioning works. Will we merge it with this?

File: scripts/prepare-release.sh [hunk 5]
```diff
@@ -2,30 +2,50 @@
 set -e
 set -o pipefail
 
-PREVIOUS_VERSION=$(grep integreatly-operator deploy/olm-catalog/integreatly-operator/integreatly-operator.package.yaml | awk -F v '{print $2}')
+if [[ -z "$TYPE" ]]; then
+  TYPE="integreatly-operator"
+fi
+
+PREVIOUS_VERSION=$(grep $TYPE deploy/olm-catalog/$TYPE/$TYPE.package.yaml | awk -F v '{print $2}') || echo "No previous version"
```
Comment (laurafitzgerald) [inline scripts/prepare-release.sh:9]: could we use jq here to pick out the currentCSV version? if there were any other instances of the $TYPE entry in the file this would pick up on it as well.

Something like

`yq r <file> channels[0].currentCSV`

I'm not super familiar with the manifests in integreatly-operator but I'm pretty sure we don't ever include historical channels so could fairly safely assume it's the first entry.

File: scripts/prepare-release.sh [hunk 6]
```diff
@@ -2,30 +2,50 @@
...
 
 update_csv() {
-  operator-sdk generate csv --csv-version "$VERSION" --default-channel --operator-name integreatly-operator --csv-channel=rhmi --update-crds
+  operator-sdk generate csv --csv-version "$VERSION" --default-channel --operator-name "$TYPE" --csv-channel=rhmi --update-crds
 }
 
 set_version() {
-  "${SED_INLINE[@]}" "s/$PREVIOUS_VERSION/$VERSION/g" Makefile
-  "${SED_INLINE[@]}" "s/$PREVIOUS_VERSION/$VERSION/g" version/version.go
+  if [[ -z "$PREVIOUS_VERSION" ]]
```
Comment (laurafitzgerald) [inline scripts/prepare-release.sh:31]: Looks like we need `$PREVIOUS_VERSION` to be set earlier in the script too. Wdyt about having that as an earlier check and not continuing unless it's set?

File: scripts/prepare-release.sh [hunk 7]
```diff
@@ -2,30 +2,50 @@
...
+  ORG="$ORG"
+fi
 
 create_new_csv() {
-  operator-sdk generate csv --csv-version "$VERSION" --default-channel --operator-name integreatly-operator --csv-channel=rhmi --update-crds --from-version "$PREVIOUS_VERSION"
+  if [[ -z "$PREVIOUS_VERSION" ]]
+    then
+      operator-sdk generate csv --csv-version "$VERSION" --default-channel --operator-name "$TYPE" --csv-channel=rhmi --update-crds
+    else
+      operator-sdk generate csv --csv-version "$VERSION" --default-channel --operator-name "$TYPE" --csv-channel=rhmi --update-crds --from-version "$PREVIOUS_VERSION"
```
Comment (laurafitzgerald) [inline scripts/prepare-release.sh:22]: I'm not sure but I think we are checking if there isn't an entry in $PREVIOUS_VERSION but then using it?

File: scripts/prepare-release.sh [hunk 8]
```diff
@@ -2,30 +2,50 @@
 set -e
 set -o pipefail
 
-PREVIOUS_VERSION=$(grep integreatly-operator deploy/olm-catalog/integreatly-operator/integreatly-operator.package.yaml | awk -F v '{print $2}')
+if [[ -z "$TYPE" ]]; then
+  TYPE="integreatly-operator"
```
Comment (laurafitzgerald) [inline scripts/prepare-release.sh:6]: Should we be really specific here? i.e. if there's no entry default to integreatly-operator but if there is entry check that it is either integreatly-operator _or_ managed-api-service?

File: deploy/olm-catalog/managed-api-service/managed-api.package-service.yaml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+channels:
```
Comment (laurafitzgerald) [inline deploy/olm-catalog/managed-api-service/managed-api.package-service.yaml:1]: Should the file name here be `managed-api-service.package.yaml`?

Comment (laurafitzgerald) [reply deploy/olm-catalog/managed-api-service/managed-api.package-service.yaml:1]: nvm. I see the comments now. I see that the directories etc should be renamed.

File: scripts/README.md (hunk 2 above)
Comment (obrienrobert) [reply scripts/README.md]: I would reword this section: `The org to where any images maybe stored on quay.io`. Maybe something like `The quay.io org to where the images will be pushed` (or something to that effect).

File: deploy/olm-catalog/managed-api-service/managed-api.package-service.yaml (hunk 4 above)
Comment (Boomatang) [reply deploy/olm-catalog/managed-api-service/managed-api.package-service.yaml:2]: The plan was to have v0.0.0 there as a place holder till we were ready to do a release. Sense you are the second person to ask this maybe we should readdress this plan.

File: scripts/prepare-release.sh (hunk 5 above)
Comment (Boomatang) [reply scripts/prepare-release.sh:9]: This is just an update to the initial script to allow for parameters and didn't see the need to rewrite the whole thing.

On a personal I don't like the use of `yq` as its not a stand tool on most systems. But that is my personal opinion and would be will to change.

File: scripts/prepare-release.sh (hunk 7 above)
Comment (Boomatang) [reply scripts/prepare-release.sh:22]: The `-z` is checking if a string has a length of zero, so we are only using the `PREVIOUS_VERSION` in the `else` which means it has been set.

File: scripts/prepare-release.sh (hunk 6 above)
Comment (Boomatang) [reply scripts/prepare-release.sh:31]: If we set this check to exit the script earlier it would remove the possibility of creating the initial CSV files. These initial CSV file would be require been created in a manual process.

File: scripts/prepare-release.sh (hunk 8 above)
Comment (Boomatang) [reply scripts/prepare-release.sh:6]: I am not sure I fully understand the question been asked but here is what I think is the answer. 

The integreatly-operator is set as the default for now for two reasons.
1. The integreatly-operator still takes priority at least till the 2.7.1 has been released
2. There has been no release of the managed-api-service to date. I feel it would be hard set the default for something not existing yet

The question about checking for  integreatly-operator or managed-api-service, I feel where it would be a good idea there would be more work required which might be pushing the scope of the ticket a bit. It might be better to have sane defaults than trying to check every possibility.

File: scripts/README.md (hunk 2 above)
Comment (Boomatang) [reply scripts/README.md]: I have done these changes locally, there maybe be other changes need to the script part. But yes `maybe` might not have been the right word to use.

**Comment by Boomatang** [comment]
I know this is after getting a bit messy. I have push changes which should allow for the work done in #1210. 

The Makefile has a list of TODO marks. These markers are for function which I feel need to be changed to allow us to have different release types. If some one who has dealt with delorean could give a once over and let me know that yes the makred function need new targets.

There is also a question should the new targets be prefixed with rhoam or try place them in existing sections with a different naming format.?

@philbrookes @laurafitzgerald @pmccarthy  calling ye out as we spoke about this earlier. 
When this is ready for a full review again I will tidy up the commits :)

File: Makefile [hunk 10]
```diff
@@ -1,20 +1,22 @@
 include ./make/*.mk
 
 ORG ?= integreatly
-NAMESPACE_PREFIX=redhat-rhmi-
+NAMESPACE_PREFIX ?= redhat-rhmi-
```
Comment (laurafitzgerald) [inline Makefile:4]: is it too early to move this namespace prefix into the block below that manages the envs for each installation. @CathalOConnorRH what did we land on in the end for the managed api prefix?

File: Makefile [hunk 11]
```diff
@@ -1,20 +1,22 @@
...
 NAMESPACE=$(NAMESPACE_PREFIX)operator
 PROJECT=integreatly-operator
+RHOAM_PROJECT=managed-api-service
```
Comment (laurafitzgerald) [inline Makefile:7]: I guess this is the project in quay? can we use one PROJECT env and move it to the block below that manages the envs?

File: Makefile [hunk 12]
```diff
@@ -1,20 +1,22 @@
...
 REG=quay.io
 SHELL=/bin/bash
-TAG ?= 2.7.0
+RHMI_TAG ?= 2.7.0
```
Comment (laurafitzgerald) [inline Makefile:10]: Same with the tags?

File: Makefile [hunk 13]
```diff
@@ -331,6 +355,7 @@ deploy/integreatly-rhmi-cr.yml:
 	sed "s/USE_CLUSTER_STORAGE/$(USE_CLUSTER_STORAGE)/g" > deploy/integreatly-rhmi-cr.yml
 	@-oc create -f deploy/integreatly-rhmi-cr.yml
 
+#TODO: new target required for rhoam
```
Comment (laurafitzgerald) [inline Makefile:358]: I think this todo is out of date. Am I right in saying we wont need to add a new target but we'll update delorean to have a new flag which we can set the installatin type by?

File: Makefile (hunk 10 above)
Comment (Boomatang) [reply Makefile:4]: I will check with @CathalOConnorRH  again but from what I understood we are not ready to make this a full time change just. I think there is one thing missing but I will follow up again.

File: Makefile (hunk 11 above)
Comment (Boomatang) [reply Makefile:7]: Yes we can move it

File: Makefile (hunk 12 above)
Comment (Boomatang) [reply Makefile:10]: The tags can be moved but the naming would want to be different. The prepare-release.sh script changes the tag values when making a new release.

File: Makefile (hunk 13 above)
Comment (Boomatang) [reply Makefile:358]: I was nopeing to push this work off to a different ticket. Currently progressing the changes needed. So yes this todo is now out of date and is been worked on.

File: Makefile (hunk 10 above)
Comment (CathalOConnorRH) [reply Makefile:4]: I think it's too early for moving this yet, I think we were waiting on RC2 but I might have missed if thats been released. I think we should wait for this release to be in prod before we change the default

Comment (Boomatang) [reply Makefile:4]: Ok @CathalOConnorRH I may have miss understood you. You were waiting to for RC2 before making the change for the default. I have added in a block which sets you the defaults depending on the install type. We can set the `NAMESPACE_PREFIX` in there. line 27 to 70 of this file. Not sure now to get a url link to work correctly. If the operator can run with the new name spaces we can do that change in the defaults block that I added.

File: Makefile (hunk 13 above)
Comment (Boomatang) [reply Makefile]: Waiting on this PR https://github.com/integr8ly/delorean/pull/148 to be merged, bu the change has been put in place

File: Makefile [hunk 14]
```diff
@@ -334,20 +360,12 @@ deploy/integreatly-rhmi-cr.yml:
...
 prepare-patch-release:
 	$(CONTAINER_ENGINE) pull quay.io/integreatly/delorean-cli:master
-	$(CONTAINER_ENGINE) run --rm -e KUBECONFIG=/kube.config -v "${HOME}/.kube/config":/kube.config:z -v "${HOME}/.delorean.yaml:/.delorean.yaml" quay.io/integreatly/delorean-cli:master delorean release openshift-ci-release --config /.delorean.yaml --version $(TAG)
+	$(CONTAINER_ENGINE) run --rm -e KUBECONFIG=/kube.config -v "${HOME}/.kube/config":/kube.config:z -v "${HOME}/.delorean.yaml:/.delorean.yaml" quay.io/integreatly/delorean-cli:master delorean release openshift-ci-release --config /.delorean.yaml --olmType $(olmType) --version $(TAG)
 
 .PHONY: release/prepare
 release/prepare:
 	@./scripts/prepare-release.sh
 
-.PHONY: push/csv
```
Comment (laurafitzgerald) [inline Makefile:343]: I've been using this recently. I think there was some discussion that we don't use it anymore. Is there an alternative way? If not can we leave this in?

File: scripts/prepare-release.sh [hunk 15]
```diff
@@ -2,30 +2,75 @@
...
+    then
+      operator-sdk generate csv --csv-version "$VERSION" --default-channel --operator-name "$OLM_TYPE" --csv-channel=rhmi --update-crds --make-manifests=false
+    else
+      operator-sdk generate csv --csv-version "$VERSION" --default-channel --operator-name "$OLM_TYPE" --csv-channel=rhmi --update-crds --from-version "$PREVIOUS_VERSION" --make-manifests=false
+  fi
 }
 
 update_csv() {
-  operator-sdk generate csv --csv-version "$VERSION" --default-channel --operator-name integreatly-operator --csv-channel=rhmi --update-crds --make-manifests=false
+  operator-sdk generate csv --csv-version "$VERSION" --default-channel --operator-name "$OLM_TYPE" --csv-channel=rhmi --update-crds --make-manifests=false
```
Comment (laurafitzgerald) [inline scripts/prepare-release.sh:40]: Do you know if this is run using your local operator-sdk or does it pick up the OPERATOR_SDK in the same way as Makefile?

File: Makefile [hunk 16]
```diff
@@ -334,20 +360,12 @@ deploy/integreatly-rhmi-cr.yml:
 .PHONY: prepare-patch-release
 prepare-patch-release:
 	$(CONTAINER_ENGINE) pull quay.io/integreatly/delorean-cli:master
-	$(CONTAINER_ENGINE) run --rm -e KUBECONFIG=/kube.config -v "${HOME}/.kube/config":/kube.config:z -v "${HOME}/.delorean.yaml:/.delorean.yaml" quay.io/integreatly/delorean-cli:master delorean release openshift-ci-release --config /.delorean.yaml --version $(TAG)
+	$(CONTAINER_ENGINE) run --rm -e KUBECONFIG=/kube.config -v "${HOME}/.kube/config":/kube.config:z -v "${HOME}/.delorean.yaml:/.delorean.yaml" quay.io/integreatly/delorean-cli:master delorean release openshift-ci-release --config /.delorean.yaml --olmType $(olmType) --version $(TAG)
```
[the remark this answers is not in the document]
Comment (Boomatang) [reply Makefile:363]: Good catch

File: scripts/prepare-release.sh (hunk 15 above)
Comment (Boomatang) [reply scripts/prepare-release.sh:40]: I'm going to say it uses the local operator-sdk , if the script is directly ran. I am not even sure if you were to use the `make release/prepare` command would it pick up the correct version.

File: Makefile (hunk 14 above)
Comment (Boomatang) [reply Makefile:343]: With the time constrain I have added this back in. Not sure if it was updated correctly to work with but rhmi and rhoam. It's some thing I don't use so not sure the best way to test.

File: scripts/prepare-release.sh (hunk 15 above)
Comment (laurafitzgerald) [reply scripts/prepare-release.sh:40]: It picks it up here: https://github.com/integr8ly/integreatly-operator/blob/master/Makefile#L35-L39

File: Makefile [hunk 17]
```diff
@@ -1,27 +1,26 @@
...
-INTEGREATLY_OPERATOR_IMAGE ?= $(REG)/$(ORG)/$(PROJECT):v$(TAG)
-MANAGED_API_OPERATOR_IMAGE ?= $(REG)/$(ORG)/managed-api-service:v$(MGDAPI_TAG)
+
+
 CONTAINER_ENGINE ?= docker
 TEST_RESULTS_DIR ?= "test-results"
 TEMP_SERVICEACCOUNT_NAME="rhmi-operator"
 CLUSTER_URL:=$(shell sh -c "oc cluster-info | grep -Eo 'https?://[-a-zA-Z0-9\.:]*'")
 
+RHMI_TAG ?= 2.7.0
```
Comment (laurafitzgerald) [inline Makefile:21]: I see that these get assigned to TAG in the if statement below. Could we just move them down there?

File: Makefile [hunk 18]
```diff
@@ -48,18 +47,42 @@ endif
...
+	export OLM_TYPE ?= integreatly-operator
+endif
+
+ifeq ($(INSTALLATION_TYPE), managed-api)
+	PROJECT=managed-api-service
+	TAG ?= RHOAM_TAG
+	OPERATOR_IMAGE=$(REG)/$(ORG)/$(PROJECT):v$(TAG)
+	NAMESPACE_PREFIX ?= redhat-rhoam-
+	APPLICATION_REPO ?= managed-api-service
+	export INSTALLATION_PREFIX ?= redhat-rhoam
```
Comment (laurafitzgerald) [inline Makefile:72]: This is used in the path of file to set up the service account running locally. https://github.com/integr8ly/integreatly-operator/blob/master/Makefile#L73
In this case I believe we need it to be `redhat-managed-api` unless there are plans to rename the folder.

File: Makefile [hunk 19]
```diff
@@ -48,18 +47,42 @@ endif
...
 export OPERATORS_IN_PRODUCT_NAMESPACE ?= false # e2e tests and createInstallationCR() need to be updated when default is changed
 export DELOREAN_PULL_SECRET_NAME ?= integreatly-delorean-pull-secret
 export ALERTING_EMAIL_ADDRESS = noreply-test@rhmi-redhat.com
 
+ifeq ($(INSTALLATION_TYPE), managed)
+	PROJECT=integreatly-operator
+	TAG ?= RHMI_TAG
+	OPERATOR_IMAGE=$(REG)/$(ORG)/$(PROJECT):v$(TAG)
+	NAMESPACE_PREFIX ?= redhat-rhmi-
+	APPLICATION_REPO ?= integreatly
```
Comment (laurafitzgerald) [inline Makefile:61]: APPLICATION_REPO confuses me. In the case of managed we are saying it's `integreatly` but in the case of managed-api` we are saying `managed-api-service`. I'm assuming this relates to application in QUAY? Do we use the application `integreatly` for rhmi and the application `managed-api-service`. I think it's probably fine but just want to be sure I'm understanding it correctly. Seems it's only used for operator-courier anyways right?

File: Makefile (hunk 14 above)
Comment (laurafitzgerald) [reply Makefile:357]: By the way, it's possible I only used this because of some legacy knowledge I have about installing RHMI. If there is another way that this is to be used, I'd like to learn about it. 😄

File: README.md [hunk 20]
```diff
@@ -1,5 +1,15 @@
 # Integreatly Operator
 
+---
+**NOTE**
```
Comment (laurafitzgerald) [inline README.md:4]: Could we also leave a similar note above the INSTALLATION_TYPE variable in the Makefile. I see that it's being left as managed (for now?). Just thinking that it should be clear that if you want to do anything in the makefile.

**Review by laurafitzgerald — COMMENTED** [review]
Hi @Boomatang just doing a short review here. Some questions inline.

File: Makefile (hunk 17 above)
Comment (Boomatang) [reply Makefile:21]: These tags are modified by the prepare-release script. In the `if` blocks the TAG is assigned  with the correct value. This value is used in the image url which is also set at this stage. Assigned the TAG value also meant not having to change any existing make targets that maybe using the TAG. Possible I should put in a comment about why the tags are done this way, Phil also asked the question.

File: README.md (hunk 20 above)
Comment (Boomatang) [reply README.md:4]: sure

File: Makefile (hunk 19 above)
Comment (Boomatang) [reply Makefile:61]: I think this is fine. These values have to do with the update_csv target which you commented about how you still use. This will set the defaults that should be used if its ran for the ORG of integreatly. Which I don't think it every will be.

File: Makefile (hunk 18 above)
Comment (Boomatang) [reply Makefile:72]: Yes this was something I was miss understanding and it needs to be changed. I will also check the file for the any other cases where this is been used.

File: scripts/prepare-release.sh (hunk 15 above)
Comment (Boomatang) [reply scripts/prepare-release.sh:40]: I think we can mark this as resolved. Call the script through the make target will use the correct operator-sdk while calling the script directly will used the default system version. From what I have seen this is the same for other scripts in the that scripts folder. The issue I can see is. Delorean calls that script directly, would delorean use it's version of the operator-sdk or in the pipelines does the correct version of the operator-sdk get installed?
